python/ray/util/concurrent/futures/ray_executor.py

```diff
@@ -0,0 +1,658 @@
+                    max_tasks_per_child must be >= 1 or None"
+                )
+        self.max_tasks_per_child = max_tasks_per_child
+
+        if initializer is not None:
+            runtime_env = kwargs.get("runtime_env")
+            if runtime_env is None or "working_dir" not in runtime_env:
+                raise ValueError(
+                    "working_dir must be specified in runtime_env dictionary if \
+                                 initializer function is not None so that \
+                                 it can be accessible by remote workers"
+                )
+
+        if max_workers is None:
+            max_workers = int(ray._private.state.cluster_resources()["CPU"])
+        if max_workers < 1:
+            raise ValueError(
+                f"max_workers={max_workers} as given. The argument "
+                "max_workers must be >= 1",
+            )
+        self.max_workers = max_workers
+
+        actor_pool_type_err = ValueError(
+            "actor_pool_type must be either 'random' or 'roundrobin'",
+        )
+        if not isinstance(actor_pool_type, str):
+            raise actor_pool_type_err
+        if actor_pool_type.lower() == "random":
+            self.actor_pool = _RandomActorPool(
+                num_actors=max_workers,
+                initializer=initializer,
+                initargs=initargs,
+                max_tasks_per_actor=self.max_tasks_per_child,
+            )
+        elif actor_pool_type.lower() == "roundrobin":
+            self.actor_pool = _RoundRobinActorPool(
+                num_actors=max_workers,
+                initializer=initializer,
+                initargs=initargs,
+                max_tasks_per_actor=self.max_tasks_per_child,
+            )
+        else:
+            raise actor_pool_type_err
+
+        return
+
+    def submit(self, fn: Callable[..., T], /, *args: Any, **kwargs: Any) -> Future[T]:
+        r"""
+        Submits a function to be executed in the actor pool with the given arguments.
+
+        Parameters
+        -----------
+        fn : Callable[]
+            A function to be executed in the actor pool as fn(\*args, \*\*kwargs)
+
+        Returns
+        -------
+        Future
+            A future representing the yet-to-be-resolved result of the
+            submitted task. Futures are also collected in self.futures.
+
+        Usage example:
+
+        .. testcode::
+
+            with RayExecutor() as ex:
+                future_0 = ex.submit(lambda x: x * x, 100)
+                future_1 = ex.submit(lambda x: x + x, 100)
+                result_0 = future_0.result()
+                result_1 = future_1.result()
+        """
+        self._check_shutdown_lock()
+        if self.actor_pool is None:
+            raise ValueError("actor_pool is not defined")
+
+        future = self.actor_pool.submit(partial(fn, *args, **kwargs))
+        self.futures.append(future)
```

Comment:
Where are the completed futures removed from the list? For long-running executors with many tasks, this list can grow indefinitely. Even though the actual task results may be garbage collected, the Future objects remain referenced in `self.futures` right?